// @allurereport/plugin-awesome/dist/generateTimeline.js
const DEFAULT_MIN_DURATION = 1;
const DEFAULT_TIMELINE_OPTIONS = {
    minDuration: DEFAULT_MIN_DURATION,
};
export const generateTimeline = async (writer, trs, options) => {
    const { timeline = DEFAULT_TIMELINE_OPTIONS } = options;
    const { minDuration = DEFAULT_MIN_DURATION } = timeline;
    const result = [];
    for (const test of trs) {
        const hasStart = Number.isInteger(test.start);
        const hasStop = Number.isInteger(test.stop);
        if (!hasStart || !hasStop) {
            continue;
        }
        const duration = test.duration ?? test.stop - test.start;
        if (duration < minDuration) {
            continue;
        }
        const { host, thread } = test.groupedLabels;
        if (!host?.length || !thread?.length) {
            continue;
        }
        result.push({
            id: test.id,
            historyId: test.historyId,
            name: test.name,
            status: test.status,
            hidden: test.hidden,
            host: host[0],
            thread: thread[0],
            environment: test.environment,
            start: test.start,
            duration,
        });
    }
    await writer.writeWidget("timeline.json", result);
};
