// @allurereport/plugin-awesome/dist/plugin.js
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var _AwesomePlugin_writer, _AwesomePlugin_generate;
import { createPluginSummary, } from "@allurereport/plugin-api";
import { preciseTreeLabels } from "@allurereport/plugin-api";
import { join } from "node:path";
import { applyCategoriesToTestResults, generateCategories } from "./categories.js";
import { filterEnv } from "./environments.js";
import { generateTimeline } from "./generateTimeline.js";
import { generateAllCharts, generateAttachmentsFiles, generateEnvironmentJson, generateEnvirontmentsList, generateGlobals, generateHistoryDataPoints, generateNav, generateQualityGateResults, generateStaticFiles, generateStatistic, generateTestCases, generateTestEnvGroups, generateTestResults, generateTree, generateTreeFilters, generateVariables, } from "./generators.js";
import { InMemoryReportDataWriter, ReportFileDataWriter } from "./writer.js";
export class AwesomePlugin {
    constructor(options = {}) {
        this.options = options;
        _AwesomePlugin_writer.set(this, void 0);
        _AwesomePlugin_generate.set(this, async (context, store) => {
            const { singleFile, groupBy = [], filter, appendTitlePath } = this.options ?? {};
            const categories = context.categories ?? [];
            const environmentItems = await store.metadataByKey("allure_environment");
            const reportEnvironments = await store.allEnvironments();
            const attachments = await store.allAttachments();
            const allTrs = await store.allTestResults({ includeHidden: true, filter });
            const statistics = await store.testsStatistic(filter);
            const environments = await store.allEnvironments();
            const envStatistics = new Map();
            const allTestEnvGroups = await store.allTestEnvGroups();
            const globalAttachments = await store.allGlobalAttachments();
            const globalExitCode = await store.globalExitCode();
            const globalErrors = await store.allGlobalErrors();
            const qualityGateResults = await store.qualityGateResultsByEnv();
            for (const env of environments) {
                envStatistics.set(env, await store.testsStatistic(filterEnv(env, filter)));
            }
            await generateStatistic(__classPrivateFieldGet(this, _AwesomePlugin_writer, "f"), {
                stats: statistics,
                statsByEnv: envStatistics,
                envs: environments,
            });
            await generateAllCharts(__classPrivateFieldGet(this, _AwesomePlugin_writer, "f"), store, this.options, context);
            const convertedTrs = await generateTestResults(__classPrivateFieldGet(this, _AwesomePlugin_writer, "f"), store, allTrs);
            applyCategoriesToTestResults(convertedTrs, categories);
            await generateCategories(__classPrivateFieldGet(this, _AwesomePlugin_writer, "f"), {
                tests: convertedTrs,
                categories,
                environmentCount: environments.length,
                environments,
                defaultEnvironment: "default",
                selectedEnvironmentCount: environments.length,
            });
            const hasGroupBy = groupBy.length > 0;
            await generateTimeline(__classPrivateFieldGet(this, _AwesomePlugin_writer, "f"), convertedTrs, this.options);
            const treeLabels = hasGroupBy
                ? preciseTreeLabels(groupBy, convertedTrs, ({ labels }) => labels.map(({ name }) => name))
                : [];
            await generateHistoryDataPoints(__classPrivateFieldGet(this, _AwesomePlugin_writer, "f"), store);
            await generateTestCases(__classPrivateFieldGet(this, _AwesomePlugin_writer, "f"), convertedTrs);
            await generateTree(__classPrivateFieldGet(this, _AwesomePlugin_writer, "f"), "tree.json", treeLabels, convertedTrs, { appendTitlePath });
            await generateNav(__classPrivateFieldGet(this, _AwesomePlugin_writer, "f"), convertedTrs, "nav.json");
            await generateTestEnvGroups(__classPrivateFieldGet(this, _AwesomePlugin_writer, "f"), allTestEnvGroups);
            for (const reportEnvironment of reportEnvironments) {
                const envConvertedTrs = convertedTrs.filter(({ environment }) => environment === reportEnvironment);
                await generateTree(__classPrivateFieldGet(this, _AwesomePlugin_writer, "f"), join(reportEnvironment, "tree.json"), treeLabels, envConvertedTrs, {
                    appendTitlePath,
                });
                await generateNav(__classPrivateFieldGet(this, _AwesomePlugin_writer, "f"), envConvertedTrs, join(reportEnvironment, "nav.json"));
                await generateCategories(__classPrivateFieldGet(this, _AwesomePlugin_writer, "f"), {
                    tests: envConvertedTrs,
                    categories,
                    environmentCount: 1,
                    defaultEnvironment: "default",
                    selectedEnvironmentCount: 1,
                    filename: join(reportEnvironment, "categories.json"),
                });
            }
            await generateTreeFilters(__classPrivateFieldGet(this, _AwesomePlugin_writer, "f"), convertedTrs);
            await generateEnvirontmentsList(__classPrivateFieldGet(this, _AwesomePlugin_writer, "f"), store);
            await generateVariables(__classPrivateFieldGet(this, _AwesomePlugin_writer, "f"), store);
            if (environmentItems?.length) {
                await generateEnvironmentJson(__classPrivateFieldGet(this, _AwesomePlugin_writer, "f"), environmentItems);
            }
            if (attachments?.length) {
                await generateAttachmentsFiles(__classPrivateFieldGet(this, _AwesomePlugin_writer, "f"), attachments, (id) => store.attachmentContentById(id));
            }
            await generateQualityGateResults(__classPrivateFieldGet(this, _AwesomePlugin_writer, "f"), qualityGateResults);
            await generateGlobals(__classPrivateFieldGet(this, _AwesomePlugin_writer, "f"), {
                globalAttachments,
                globalErrors,
                globalExitCode,
                contentFunction: (id) => store.attachmentContentById(id),
            });
            const reportDataFiles = singleFile ? __classPrivateFieldGet(this, _AwesomePlugin_writer, "f").reportFiles() : [];
            await generateStaticFiles({
                ...this.options,
                id: context.id,
                allureVersion: context.allureVersion,
                reportFiles: context.reportFiles,
                reportUuid: context.reportUuid,
                reportName: context.reportName,
                ci: context.ci,
                reportDataFiles,
            });
        });
        this.start = async (context) => {
            const { singleFile } = this.options;
            if (singleFile) {
                __classPrivateFieldSet(this, _AwesomePlugin_writer, new InMemoryReportDataWriter(), "f");
                return;
            }
            __classPrivateFieldSet(this, _AwesomePlugin_writer, new ReportFileDataWriter(context.reportFiles), "f");
            await Promise.resolve();
        };
        this.update = async (context, store) => {
            if (!__classPrivateFieldGet(this, _AwesomePlugin_writer, "f")) {
                throw new Error("call start first");
            }
            await __classPrivateFieldGet(this, _AwesomePlugin_generate, "f").call(this, context, store);
        };
        this.done = async (context, store) => {
            if (!__classPrivateFieldGet(this, _AwesomePlugin_writer, "f")) {
                throw new Error("call start first");
            }
            await __classPrivateFieldGet(this, _AwesomePlugin_generate, "f").call(this, context, store);
        };
    }
    async info(context, store) {
        return createPluginSummary({
            name: this.options.reportName || context.reportName,
            plugin: "Awesome",
            meta: {
                reportId: context.reportUuid,
                singleFile: this.options.singleFile ?? false,
                withTestResultsLinks: true,
            },
            filter: this.options.filter,
            ci: context.ci,
            history: context.history,
            store,
        });
    }
}
_AwesomePlugin_writer = new WeakMap(), _AwesomePlugin_generate = new WeakMap();
