// @allurereport/plugin-testops/dist/client.js
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _TestOpsClient_accessToken, _TestOpsClient_projectId, _TestOpsClient_oauthToken, _TestOpsClient_client, _TestOpsClient_launch, _TestOpsClient_session, _TestOpsClient_uploadInProgress, _TestOpsClient_uploadLimit;
import axios from "axios";
import FormData from "form-data";
import chunk from "lodash.chunk";
import pLimit from "p-limit";
export class TestOpsClient {
    constructor(params) {
        _TestOpsClient_accessToken.set(this, void 0);
        _TestOpsClient_projectId.set(this, void 0);
        _TestOpsClient_oauthToken.set(this, "");
        _TestOpsClient_client.set(this, void 0);
        _TestOpsClient_launch.set(this, void 0);
        _TestOpsClient_session.set(this, void 0);
        _TestOpsClient_uploadInProgress.set(this, false);
        _TestOpsClient_uploadLimit.set(this, 1);
        if (!params.accessToken) {
            throw new Error("accessToken is required");
        }
        if (!params.projectId) {
            throw new Error("projectId is required");
        }
        if (!params.baseUrl) {
            throw new Error("baseUrl is required");
        }
        if (params.limit && params.limit > 5) {
            throw new Error("limit can't be greater than 5");
        }
        __classPrivateFieldSet(this, _TestOpsClient_accessToken, params.accessToken, "f");
        __classPrivateFieldSet(this, _TestOpsClient_projectId, params.projectId, "f");
        __classPrivateFieldSet(this, _TestOpsClient_client, axios.create({
            baseURL: params.baseUrl,
            validateStatus: (status) => status >= 200 && status < 400,
        }), "f");
        if (params.limit) {
            __classPrivateFieldSet(this, _TestOpsClient_uploadLimit, params.limit, "f");
        }
    }
    get launchUrl() {
        if (!__classPrivateFieldGet(this, _TestOpsClient_launch, "f")) {
            return undefined;
        }
        return new URL(`launch/${__classPrivateFieldGet(this, _TestOpsClient_launch, "f").id}`, __classPrivateFieldGet(this, _TestOpsClient_client, "f").defaults.baseURL).toString();
    }
    async issueOauthToken() {
        const formData = new FormData();
        formData.append("grant_type", "apitoken");
        formData.append("scope", "openid");
        formData.append("token", __classPrivateFieldGet(this, _TestOpsClient_accessToken, "f"));
        const { data } = await __classPrivateFieldGet(this, _TestOpsClient_client, "f").post("/api/uaa/oauth/token", formData);
        __classPrivateFieldSet(this, _TestOpsClient_oauthToken, data.access_token, "f");
    }
    async startUpload(ci) {
        if (!__classPrivateFieldGet(this, _TestOpsClient_launch, "f")) {
            throw new Error("Launch isn't created! Call createLaunch first");
        }
        await __classPrivateFieldGet(this, _TestOpsClient_client, "f").post("/api/upload/start", {
            projectId: __classPrivateFieldGet(this, _TestOpsClient_projectId, "f"),
            ci: {
                name: ci.type,
            },
            job: {
                name: ci.jobUid,
                uid: ci.jobUid,
            },
            jobRun: {
                uid: ci.jobRunUid,
            },
            launch: {
                id: __classPrivateFieldGet(this, _TestOpsClient_launch, "f").id,
            },
        }, {
            headers: {
                Authorization: `Bearer ${__classPrivateFieldGet(this, _TestOpsClient_oauthToken, "f")}`,
            },
        });
        __classPrivateFieldSet(this, _TestOpsClient_uploadInProgress, true, "f");
    }
    async stopUpload(ci, status) {
        if (!__classPrivateFieldGet(this, _TestOpsClient_uploadInProgress, "f")) {
            throw new Error("Upload isn't started! Call startUpload first");
        }
        await __classPrivateFieldGet(this, _TestOpsClient_client, "f").post("/api/upload/stop", {
            jobRunUid: ci.jobRunUid,
            jobUid: ci.jobUid,
            projectId: __classPrivateFieldGet(this, _TestOpsClient_projectId, "f"),
            status,
        }, {
            headers: {
                Authorization: `Bearer ${__classPrivateFieldGet(this, _TestOpsClient_oauthToken, "f")}`,
            },
        });
        __classPrivateFieldSet(this, _TestOpsClient_uploadInProgress, false, "f");
    }
    async createLaunch(launchName, launchTags) {
        const { data } = await __classPrivateFieldGet(this, _TestOpsClient_client, "f").post("/api/launch", {
            name: launchName,
            projectId: __classPrivateFieldGet(this, _TestOpsClient_projectId, "f"),
            autoclose: true,
            external: true,
            tags: launchTags.map((tag) => ({ name: tag })),
        }, {
            headers: {
                Authorization: `Bearer ${__classPrivateFieldGet(this, _TestOpsClient_oauthToken, "f")}`,
            },
        });
        __classPrivateFieldSet(this, _TestOpsClient_launch, data, "f");
    }
    async createSession(environment = {}) {
        if (!__classPrivateFieldGet(this, _TestOpsClient_launch, "f")) {
            throw new Error("Launch isn't created! Call createLaunch first");
        }
        const { data } = await __classPrivateFieldGet(this, _TestOpsClient_client, "f").post("/api/upload/session?manual=true", {
            launchId: __classPrivateFieldGet(this, _TestOpsClient_launch, "f").id,
            environment: Object.entries(environment).map(([key, value]) => ({ key, value: String(value) })),
        }, {
            headers: {
                Authorization: `Bearer ${__classPrivateFieldGet(this, _TestOpsClient_oauthToken, "f")}`,
            },
        });
        __classPrivateFieldSet(this, _TestOpsClient_session, data, "f");
    }
    async uploadTestResults(params) {
        if (!__classPrivateFieldGet(this, _TestOpsClient_session, "f")) {
            throw new Error("Session isn't created! Call createSession first");
        }
        const { trs, attachmentsResolver, fixturesResolver, onProgress } = params;
        const trsChunks = chunk(trs, 100);
        const uploadLimitFn = pLimit(__classPrivateFieldGet(this, _TestOpsClient_uploadLimit, "f"));
        await Promise.all(trsChunks.map(async (trsChunk) => {
            const { data } = await __classPrivateFieldGet(this, _TestOpsClient_client, "f").post("/api/upload/test-result", {
                testSessionId: __classPrivateFieldGet(this, _TestOpsClient_session, "f").id,
                results: trsChunk.map((tr) => ({
                    ...tr,
                    uuid: tr.id,
                })),
            }, {
                headers: {
                    "Authorization": `Bearer ${__classPrivateFieldGet(this, _TestOpsClient_oauthToken, "f")}`,
                    "Content-Type": "application/json",
                },
            });
            const trsTestOpsIdsByUuid = data.results.reduce((acc, { id, uuid }) => ({ ...acc, [uuid]: id }), {});
            await Promise.all(trsChunk.map((tr) => uploadLimitFn(async () => {
                const trTestOpsId = trsTestOpsIdsByUuid[tr.id];
                const attachments = await attachmentsResolver(tr);
                const fixtures = await fixturesResolver(tr);
                if (attachments.length > 0) {
                    const attachmentsChunks = chunk(attachments, 100);
                    await Promise.all(attachmentsChunks.map(async (attachmentsChunk) => {
                        const formData = new FormData();
                        attachmentsChunk.forEach((attachment) => {
                            formData.append("file", attachment.content, {
                                filename: attachment.originalFileName,
                                contentType: attachment.contentType,
                            });
                        });
                        await __classPrivateFieldGet(this, _TestOpsClient_client, "f").post(`/api/upload/test-result/${trTestOpsId}/attachment`, formData, {
                            headers: {
                                Authorization: `Bearer ${__classPrivateFieldGet(this, _TestOpsClient_oauthToken, "f")}`,
                                ...formData.getHeaders(),
                            },
                        });
                    }));
                }
                if (fixtures.length > 0) {
                    await __classPrivateFieldGet(this, _TestOpsClient_client, "f").post(`/api/upload/test-result/${trTestOpsId}/test-fixture-result`, {
                        fixtures,
                    }, {
                        headers: {
                            Authorization: `Bearer ${__classPrivateFieldGet(this, _TestOpsClient_oauthToken, "f")}`,
                        },
                    });
                }
                onProgress?.();
            })));
        }));
    }
}
_TestOpsClient_accessToken = new WeakMap(), _TestOpsClient_projectId = new WeakMap(), _TestOpsClient_oauthToken = new WeakMap(), _TestOpsClient_client = new WeakMap(), _TestOpsClient_launch = new WeakMap(), _TestOpsClient_session = new WeakMap(), _TestOpsClient_uploadInProgress = new WeakMap(), _TestOpsClient_uploadLimit = new WeakMap();
