// @allurereport/plugin-testops/dist/plugin.js
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _TestopsPlugin_instances, _TestopsPlugin_ci, _TestopsPlugin_client, _TestopsPlugin_launchName, _TestopsPlugin_launchTags, _TestopsPlugin_uploadedTestResultsIds, _TestopsPlugin_upload, _TestopsPlugin_startUpload, _TestopsPlugin_stopUpload;
import { detect } from "@allurereport/ci";
import { getWorstStatus } from "@allurereport/core-api";
import { createPluginSummary, } from "@allurereport/plugin-api";
import { env } from "node:process";
import ProgressBar from "progress";
import { TestOpsClient } from "./client.js";
import { resolvePluginOptions, unwrapStepsAttachments } from "./utils.js";
export class TestopsPlugin {
    constructor(options) {
        _TestopsPlugin_instances.add(this);
        this.options = options;
        _TestopsPlugin_ci.set(this, void 0);
        _TestopsPlugin_client.set(this, void 0);
        _TestopsPlugin_launchName.set(this, "");
        _TestopsPlugin_launchTags.set(this, []);
        _TestopsPlugin_uploadedTestResultsIds.set(this, []);
        const { accessToken, endpoint, projectId, launchName, launchTags } = resolvePluginOptions(options);
        __classPrivateFieldSet(this, _TestopsPlugin_ci, detect(), "f");
        if ([accessToken, endpoint, projectId].every(Boolean)) {
            __classPrivateFieldSet(this, _TestopsPlugin_client, new TestOpsClient({
                baseUrl: endpoint,
                accessToken,
                projectId,
            }), "f");
            __classPrivateFieldSet(this, _TestopsPlugin_launchName, launchName, "f");
            __classPrivateFieldSet(this, _TestopsPlugin_launchTags, launchTags, "f");
        }
        if (!accessToken) {
            console.warn("TestOps access token is missing. Please provide a valid access token in the plugin options.");
        }
        if (!endpoint) {
            console.warn("TestOps endpoint is missing. Please provide a valid endpoint in the plugin options.");
        }
        if (!projectId) {
            console.warn("TestOps project ID is missing. Please provide a valid project ID in the plugin options.");
        }
    }
    get ciMode() {
        return __classPrivateFieldGet(this, _TestopsPlugin_ci, "f") && __classPrivateFieldGet(this, _TestopsPlugin_ci, "f").type !== "local";
    }
    async start(_context, store) {
        if (!__classPrivateFieldGet(this, _TestopsPlugin_client, "f")) {
            return;
        }
        await __classPrivateFieldGet(this, _TestopsPlugin_instances, "m", _TestopsPlugin_startUpload).call(this);
        await __classPrivateFieldGet(this, _TestopsPlugin_instances, "m", _TestopsPlugin_upload).call(this, store, { issueNewToken: false });
        console.info(`TestOps launch has been created: ${__classPrivateFieldGet(this, _TestopsPlugin_client, "f").launchUrl}`);
    }
    async update(_context, store) {
        if (!__classPrivateFieldGet(this, _TestopsPlugin_client, "f")) {
            return;
        }
        await __classPrivateFieldGet(this, _TestopsPlugin_instances, "m", _TestopsPlugin_upload).call(this, store);
    }
    async done(_context, store) {
        if (!__classPrivateFieldGet(this, _TestopsPlugin_client, "f")) {
            return;
        }
        const allTrs = (await store.allTestResults()).filter((tr) => this.options.filter ? this.options.filter(tr) : true);
        const worstStatus = getWorstStatus(allTrs.map(({ status }) => status));
        await __classPrivateFieldGet(this, _TestopsPlugin_instances, "m", _TestopsPlugin_upload).call(this, store);
        await __classPrivateFieldGet(this, _TestopsPlugin_instances, "m", _TestopsPlugin_stopUpload).call(this, worstStatus || "unknown");
    }
    async info(context, store) {
        if (!__classPrivateFieldGet(this, _TestopsPlugin_client, "f")?.launchUrl) {
            return undefined;
        }
        return {
            ...(await createPluginSummary({
                name: __classPrivateFieldGet(this, _TestopsPlugin_launchName, "f"),
                plugin: "TestOps",
                meta: {
                    reportId: context.reportUuid,
                },
                filter: this.options.filter,
                history: context.history,
                ci: context.ci,
                store,
            })),
            remoteHref: __classPrivateFieldGet(this, _TestopsPlugin_client, "f").launchUrl,
        };
    }
}
_TestopsPlugin_ci = new WeakMap(), _TestopsPlugin_client = new WeakMap(), _TestopsPlugin_launchName = new WeakMap(), _TestopsPlugin_launchTags = new WeakMap(), _TestopsPlugin_uploadedTestResultsIds = new WeakMap(), _TestopsPlugin_instances = new WeakSet(), _TestopsPlugin_upload = async function _TestopsPlugin_upload(store, options) {
    const { issueNewToken = true } = options ?? {};
    const allTrs = await store.allTestResults();
    const trsToUpload = allTrs.filter((tr) => {
        const uploaded = __classPrivateFieldGet(this, _TestopsPlugin_uploadedTestResultsIds, "f").includes(tr.id);
        if (this.options.filter) {
            return this.options.filter(tr) && !uploaded;
        }
        return !uploaded;
    });
    if (trsToUpload.length === 0) {
        return;
    }
    const allTrsWithAttachments = trsToUpload.map((tr) => {
        return {
            ...tr,
            steps: unwrapStepsAttachments(tr.steps),
        };
    });
    if (issueNewToken) {
        await __classPrivateFieldGet(this, _TestopsPlugin_client, "f").issueOauthToken();
    }
    const progressBar = new ProgressBar("Uploading test results [:bar] :current/:total", {
        total: allTrsWithAttachments.length,
        width: 20,
    });
    progressBar.render();
    await __classPrivateFieldGet(this, _TestopsPlugin_client, "f").createSession(env);
    await __classPrivateFieldGet(this, _TestopsPlugin_client, "f").uploadTestResults({
        trs: allTrsWithAttachments,
        onProgress: () => progressBar.tick(),
        attachmentsResolver: async (tr) => {
            const attachments = await store.attachmentsByTrId(tr.id);
            return await Promise.all(attachments.map(async (attachment) => {
                const content = await store.attachmentContentById(attachment.id);
                return {
                    originalFileName: attachment.originalFileName,
                    contentType: attachment.contentType,
                    content: await content?.readContent(async (s) => s),
                };
            }));
        },
        fixturesResolver: async (tr) => {
            const fxts = await store.fixturesByTrId(tr.id);
            return fxts.map((fxt) => ({
                ...fxt,
                type: fxt.type.toUpperCase(),
                steps: unwrapStepsAttachments(fxt.steps),
            }));
        },
    });
    __classPrivateFieldGet(this, _TestopsPlugin_uploadedTestResultsIds, "f").push(...allTrsWithAttachments.map((tr) => tr.id));
}, _TestopsPlugin_startUpload = async function _TestopsPlugin_startUpload() {
    if (!__classPrivateFieldGet(this, _TestopsPlugin_client, "f")) {
        return;
    }
    await __classPrivateFieldGet(this, _TestopsPlugin_client, "f").issueOauthToken();
    await __classPrivateFieldGet(this, _TestopsPlugin_client, "f").createLaunch(__classPrivateFieldGet(this, _TestopsPlugin_launchName, "f"), __classPrivateFieldGet(this, _TestopsPlugin_launchTags, "f"));
    if (!this.ciMode) {
        return;
    }
    await __classPrivateFieldGet(this, _TestopsPlugin_client, "f").startUpload(__classPrivateFieldGet(this, _TestopsPlugin_ci, "f"));
}, _TestopsPlugin_stopUpload = async function _TestopsPlugin_stopUpload(status) {
    if (!this.ciMode || !__classPrivateFieldGet(this, _TestopsPlugin_client, "f")) {
        return;
    }
    await __classPrivateFieldGet(this, _TestopsPlugin_client, "f").stopUpload(__classPrivateFieldGet(this, _TestopsPlugin_ci, "f"), status);
};
